File: android/src/main/java/me/carda/awesome_notifications/core/utils/TimeZoneUtils.java
package me.carda.awesome_notifications.core.utils;

import java.util.TimeZone;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.annotation.Nullable;

public class TimeZoneUtils {

    // *****************************  SINGLETON PATTERN  *****************************

    protected static final TimeZoneUtils instance = new TimeZoneUtils();

    private TimeZoneUtils(){}

    public static TimeZoneUtils getInstance(){
        return instance;
    }

    // *******************************************************************************

    @Nullable
    public TimeZone getValidTimeZone(@Nullable String timeZoneId){
        if(timeZoneId == null)
            return null;

        final String regex = "^((\\-|\\+)?(\\d{2}(:\\d{2})?))|(\\S+)$";

        final Pattern pattern = Pattern.compile(regex, Pattern.MULTILINE);
        final Matcher matcher = pattern.matcher(timeZoneId);

        TimeZone finalTimeZone;
        if(matcher.matches()){
            if(matcher.group(1) != null) {
                finalTimeZone = TimeZone.getTimeZone(
                        "GMT"+
                                (matcher.group(2) == null ? "+" : matcher.group(2)) +
                                matcher.group(3) +
                                (matcher.group(4) == null ? ":00" : matcher.group(4)));
            }
            else {
                finalTimeZone = TimeZone.getTimeZone(matcher.group(5));
            }
        }
        else {
            finalTimeZone = TimeZone.getDefault();
        }

        if(finalTimeZone.getID().contains(timeZoneId))
            return finalTimeZone;

        return null;
    }

    @Nullable
    public String timeZoneToString(@Nullable TimeZone timeZone){
        if(timeZone == null)
            return null;

        return timeZone.getID();
    }
}

File: ios/AwnCore/Classes/utils/TimeZoneUtils.swift
//
//  TimeZoneUtils.swift
//  awesome_notifications
//
//  Created by CardaDev on 09/03/22.
//

import Foundation

public class TimeZoneUtils {

    // *****************************  SINGLETON PATTERN  *****************************
    
    static var instance:TimeZoneUtils?
    public static var shared:TimeZoneUtils {
        get {
            TimeZoneUtils.instance = TimeZoneUtils.instance ?? TimeZoneUtils()
            return TimeZoneUtils.instance!
        }
    }
    private init(){}

    // *******************************************************************************

    public func getValidTimeZone(fromTimeZoneId timeZoneId:String?) -> TimeZone? {
        guard let timeZoneId:String = timeZoneId
        else {
            return nil
        }
            
        let pattern = #"^((\-|\+)?(\d{2}(:\d{2})?))|(\S+)$"#
        let matcher = timeZoneId.matchList(pattern)
        
        var finalTimeZone:TimeZone?
        if !matcher.isEmpty {
            let matches:[String] = matcher[0]
            if !matches[1].isEmpty {
                let finalTimeZoneId:String =
                "GMT"+(
                    (matches[2].isEmpty ? "+" : matches[2]) +
                    matches[3] +
                    (matches[4].isEmpty ? ":00" : "")
                )
                finalTimeZone = TimeZone.init(identifier:finalTimeZoneId)
            }
            else {
              finalTimeZone = TimeZone.init(identifier: matches[5])
            }
        }
        
        if finalTimeZone?.identifier.contains(timeZoneId) ?? false {
            return finalTimeZone
        }
        
        return nil
    }

    public func timeZoneToString(timeZone:TimeZone?) -> String? {
        if timeZone == nil {
            return nil
        }
        else {
            return timeZone!.identifier
        }
    }
}
